inference: NFC refactoring on `Conditional`/`InterConditional`

aviatesk opened the pull request:
When I introduced this new lattice within #38905, I defined this in
`Core` for some performance reason, but now I'm sure it is really
unnecessary. Rather, it is easier to develop Julia-level inference routine
in pure Julia, so the first commit commit moves the definition of 
`InterConditional` to `Core.Compiler`.

The second commit renames `.vtype` field of `Conditional` to `.thentype`.
The rational here is that the "vtype" term is a bit confusing since it is
used for "variable state" managed by `InferenceState`. Rather, the `.thentype`
naming is more consistent with the counterpart field `.elsetype`, and 
would make its meaning more explicit.

Finally, the third commit makes them more symmetic by renaming the 
`.var::SlotNumber` field of `Conditional` to `.slot::Int` as like 
`InterConditional`. This change may also cut off few allocations of `SlotNumber`.
We may want `Conditional` to propagate constraints about non-local binding
(like global references) in the future, but I think that won't happen that
soonish. For the meanwhile, it would be reasonable to keep the field
specific `SlotNumber`.

NOTE: I'd like to merge these commits separately.

@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk committed:
inference: define `InterConditional` within `Core.Compiler`

When I introduced this new lattice within #38905, I defined this in
`Core` for some performance reason, but now I'm sure it is really
necessary. Rather, it is easier to develop Julia-level inference routine
in pure Julia, so this commit moves the definition of `InterConditional`
to `Core.Compiler`.

aviatesk committed:
inference: rename `.vtype` of `Conditional` to `.thentype`

The "vtype" term is a bit confusing since it is used for "variable state"
managed by `InferenceState`.
Rather, the `.thentype` naming is more consistent with the counterpart
field `.elsetype`, and would make its meaning more explicit.

aviatesk committed:
inference: make `Conditional` and `InterConditional` slightly more symmetric

Especially the `.var::SlotNumber` field of `Conditional` is now renamed
to `.slot::Int` as like `InterConditional`. This change may also cut off
few allocations of `SlotNumber`.

We may want `Conditional` to propagate constraints about non-local binding
(like global references) in the future, but I think that won't happen that
soonish. For the meanwhile, it would be reasonable to keep the field
specific `SlotNumber`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45479) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f6ef261_vs_84e9989/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45479#issuecomment-1139488520) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d421882_vs_84e9989/report.md).

vtjnash commented:
We need to define these in Core so that Revise works on the Compiler. Otherwise, it gets confused and crashes Julia when loaded.

aviatesk commented:
Hmm, I just confirmed `Revise.track(Core.Compiler)` works as expected on this branch. Seems like we no longer need that restriction?

aviatesk commented:
In a case when it will cause a problem to define `InterConditional` in `Core.Compiler`, I rebased the commits and moved 75ffa0aaf8a33be3368aaa1eb3f72a66344efd03 to the last so that we can easily revert it. I'm planning to merge them separately once I confirm CI finishes successfully.

vtjnash commented:
Crashing is not really a simple thing you check for by doing one test. The previous behavior observed is that if something tries to revise the wrong part of `Core.Compiler`, it will cause other parts to see the old data in the cache and get confused and crash. Sometimes people make plugins also that expect consistent definitions of all of the types that are exported by the Compiler module interfaces. Putting these types into `Core` had been avoiding problems with that cache becoming invalid.
